### ISSUE TITLE
useless fallthrought: report them

### ISSUE THREAD

**Issue by ccoVeille** [mgechev/revive#1450]
Here it would be interesting to consider reporting empty fallthrought

```go
switch a {
  case 1:
    bar()
	case 2:
    fallthrought
  case 3:
		foo()
}
```

Maybe with an other PR anyway, and maybe another rule

But this would be expected

```go
switch a {
  case 1:
    bar()
	case 2, 3:
		foo()
}
```

_Originally posted by @ccoVeille in https://github.com/mgechev/revive/pull/1448#discussion_r2241802977_

**Comment by chavacava** [comment]
I ran a draft version of the rule on some code bases and maybe the rule should **not** warn when falling-through the `default` clause.
These clauses falling through to `default` seem to be used as placeholders/documentation.

For example:

```go
	switch kind {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		value := valueHelp.asInt64(val)
		return MaximumInt(path, in, value, int64(max), exclusive)
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		value := valueHelp.asUint64(val)
		if max < 0 {
			return errors.ExceedsMaximum(path, in, max, exclusive, val)
		}
		return MaximumUint(path, in, value, uint64(max), exclusive)
	case reflect.Float32, reflect.Float64:
		fallthrough
	default:
		value := valueHelp.asFloat64(val)
		return Maximum(path, in, value, max, exclusive)
	}
```

What do you say?

**Comment by ccoVeille** [comment]
I would report them with a lower confidence.

**Comment by ccoVeille** [comment]
Did you find real life strange thing that the rule could help to fix ?

**Comment by chavacava** [comment]
> Did you find real life strange thing that the rule could help to fix ?

I think so.

From things like

```go
	switch subjectKind {
	case rbacv1.GroupKind:
		sar = &authv1.SubjectAccessReview{
			Spec: authv1.SubjectAccessReviewSpec{
				Groups:             []string{subject},
				ResourceAttributes: resource,
			},
		}
	case rbacv1.UserKind:
		fallthrough
	case rbacv1.ServiceAccountKind:
		sar = &authv1.SubjectAccessReview{
			Spec: authv1.SubjectAccessReviewSpec{
				User:               subject,
				ResourceAttributes: resource,
			},
		}
	default:
		framework.Failf("invalid subjectKind %s", subjectKind)
	}
``` 

to 

```go
	switch n.Type {
	case Document:
		fallthrough
	case BlockQuote:
		fallthrough
	case List:
		fallthrough
	case Item:
		fallthrough
	case Paragraph:
		fallthrough
	case Heading:
		fallthrough
	case Emph:
		fallthrough
	case Strong:
		fallthrough
	case Del:
		fallthrough
	case Link:
		fallthrough
	case Image:
		fallthrough
	case Table:
		fallthrough
	case TableHead:
		fallthrough
	case TableBody:
		fallthrough
	case TableRow:
		fallthrough
	case TableCell:
		return true
	default:
		return false
	}
```

**Comment by chavacava** [comment]
Side note on performance. As I suspected, using useless `fallthrough` is suboptimal.

For
```go
func toto(a int) bool {
	switch a {
	case 0:
		fallthrough
	case 1:
		fallthrough
	case 2:
		return true
	default:
		return false
	}
}
```
the resulting SSA is
```
# /app/public/buildbox/e2419dfd-7102-11f0-9e73-0242ac16000d/main.go
00000 (3) TEXT command-line-arguments.toto(SB), ABIInternal
00001 (3) FUNCDATA $0, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
00002 (3) FUNCDATA $1, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
00003 (3) FUNCDATA $5, command-line-arguments.toto.arginfo1(SB)
00004 (3) FUNCDATA $6, command-line-arguments.toto.argliveinfo(SB)
b1
00005 (3) PCDATA $3, $1
v21
00006 (+5) TESTQ AX, AX
b1
00007 (5) JEQ 12
v5
00008 (+7) CMPQ AX, $1
b5
00009 (7) JEQ 12
v4
00010 (+9) CMPQ AX, $2
b9
00011 (9) JNE 14
v22
00012 (+10) MOVL $1, AX
b13
00013 (10) RET
v23
00014 (+12) XORL AX, AX
b14
00015 (12) RET
00016 (?) END
```
while the SSA for the switch with consolidated case clauses 
```go
func toto(a int) bool {
	switch a {
	case 0, 1, 2:
		return true
	default:
		return false
	}
}
```
is

```
# /app/public/buildbox/06de37da-7103-11f0-9e73-0242ac16000d/main.go
00000 (3) TEXT command-line-arguments.toto(SB), ABIInternal
00001 (3) FUNCDATA $0, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
00002 (3) FUNCDATA $1, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
00003 (3) FUNCDATA $5, command-line-arguments.toto.arginfo1(SB)
00004 (3) FUNCDATA $6, command-line-arguments.toto.argliveinfo(SB)
b5
00005 (3) PCDATA $3, $1
v3
00006 (+5) CMPQ AX, $2
b5
00007 (5) JHI 10
v20
00008 (+6) MOVL $1, AX
b6
00009 (6) RET
v19
00010 (+8) XORL AX, AX
b7
00011 (8) RET
00012 (?) END
```

**Comment by ccoVeille** [comment]
Thanks

**Comment by veeam-denis** [comment]
@chavacava I'm curious: how did you get these assembly listings?

**Comment by chavacava** [comment]
> @chavacava I'm curious: how did you get these assembly listings?

I used to use the Go toolchain to generate a web page with info on all compilation phases but I forgot the command :(

For the above examples I've used https://golang.design/gossa
